src/renderers/dom/shared/__tests__/ReactServerRendering-test.js

```diff
@@ -23,20 +23,132 @@ var ReactTestUtils;
+  return renderIntoDom(element, div, errorCount);
+};
+
+const clientRenderOnServerString = (element, errorCount = 0) => {
+  return serverRender(element, errorCount).then((markup) => {
+    resetModules();
+    var domElement = document.createElement('div');
+    domElement.innerHTML = markup;
+    return renderIntoDom(element, domElement, errorCount);
+  });
+};
+
+const clientRenderOnBadMarkup = (element, errorCount = 0) => {
+  var domElement = document.createElement('div');
+  domElement.innerHTML = '<div id="badIdWhichWillCauseMismatch" data-reactroot="" data-reactid="1"></div>';
+  return renderIntoDom(element, domElement, errorCount + 1);
+};
+
+// runs a DOM rendering test as four different tests, with four different rendering
+// scenarios:
+// -- render to string on server
+// -- render on client without any server markup "clean client render"
+// -- render on client on top of good server-generated string markup
+// -- render on client on top of bad server-generated markup
+//
+// testFn is a test that has one arg, which is a render function. the render
+// function takes in a ReactElement and an optional expected error count and
+// returns a promise of a DOM Element.
+//
+// You should only perform tests that examine the DOM of the results of
+// render; you should not depend on the interactivity of the returned DOM element,
+// as that will not work in the server string scenario.
+function itRenders(desc, testFn) {
+  it(`${desc} with server string render`,
+    () => testFn(serverRender));
+  itClientRenders(desc, testFn);
+}
+
+// run testFn in three different rendering scenarios:
+// -- render on client without any server markup "clean client render"
+// -- render on client on top of good server-generated string markup
+// -- render on client on top of bad server-generated markup
+//
+// testFn is a test that has one arg, which is a render function. the render
+// function takes in a ReactElement and an optional expected error count and
+// returns a promise of a DOM Element.
+//
+// Since all of the renders in this function are on the client, you can test interactivity,
+// unlike with itRenders.
+function itClientRenders(desc, testFn) {
+  it(`${desc} with clean client render`,
+    () => testFn(clientCleanRender));
+  it(`${desc} with client render on top of good server markup`,
+    () => testFn(clientRenderOnServerString));
+  it(`${desc} with client render on top of bad server markup`,
+    () => testFn(clientRenderOnBadMarkup));
+}
+
+function resetModules() {
+  jest.resetModuleRegistry();
+  React = require('React');
+  ReactDOM = require('ReactDOM');
+  ReactDOMFeatureFlags = require('ReactDOMFeatureFlags');
+  ReactMarkupChecksum = require('ReactMarkupChecksum');
+  ReactTestUtils = require('ReactTestUtils');
+  ReactReconcileTransaction = require('ReactReconcileTransaction');
+
+  ExecutionEnvironment = require('ExecutionEnvironment');
+  ExecutionEnvironment.canUseDOM = false;
```

Comment:
You're right; this was an incantation I copied from the old code without thinking. I've changed it in a91c836 to change canUseDom to true before the call to `ReactDOM.render` in `renderToDom` (and change it back after). Seems to work right AFAICT.